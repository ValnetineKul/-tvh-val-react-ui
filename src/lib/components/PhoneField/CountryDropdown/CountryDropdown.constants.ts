import type { Surface } from '../../../../themes/core';

export const surface: Surface = '100';
