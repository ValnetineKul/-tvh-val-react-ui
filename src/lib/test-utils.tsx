import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
