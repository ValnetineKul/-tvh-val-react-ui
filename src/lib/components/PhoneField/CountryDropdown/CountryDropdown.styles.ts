import { makeStyles } from '../../../../themes/core';
import { surface } from './CountryDropdown.constants';

const useStyles = makeStyles({ name: 'CountryDropDown' })((theme) => ({
  root: {
    position: 'absolute',
    zIndex: 2,
    width: `calc(${theme.layoutSpacing['spacing/550']} + ${theme.layoutSpacing['spacing/300']})`,
    height: theme.layoutSpacing['spacing/550'],
    display: 'flex',
    alignItems: 'center',
  },
  select: {
    height: `calc(100% - ${theme.layoutSpacing['spacing/100']})`,
    width: `calc(100% - ${theme.layoutSpacing['spacing/100']})`,
    marginLeft: '1px',
    color: theme.color[`icon/onSurface${surface}/default` as const],

    '&:disabled': {
      backgroundColor: theme.color['bg/disabled'],
    },
  },
  selectError: {
    backgroundColor: 'transparent',
  },
  flagIcon: {
    marginLeft: theme.layoutSpacing['spacing/300'],
  },
}));

export default useStyles;
