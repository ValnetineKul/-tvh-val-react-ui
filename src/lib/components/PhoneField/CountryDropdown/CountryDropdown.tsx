import type { FC } from 'react';
import React, { useMemo, useState } from 'react';
import { useScreenSize } from '../../../hooks';
import type { CountryCode, DataAttributes } from '../../../types/common';
import DropdownList from '../../Dropdowns/DropdownList';
import MenuItem from '../../Menus/MenuItem';
import ButtonBase from '../../Surfaces/ButtonBase';
import FlagIcon from '../FlagIcon/FlagIcon';
import { surface } from './CountryDropdown.constants';
import useStyles from './CountryDropdown.styles';

type Country = { label: string; value: CountryCode };

export interface CountryDropdownProps {
  value: CountryCode;
  options: Country[];
  disabled?: boolean;
  readOnly?: boolean;
  noResultsMessage?: string;
  errorMessage?: string;
  onChange: (value: string) => void;
  buttonProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
  dropdownTitle: string;
}

const CountryDropdown: FC<CountryDropdownProps> = ({
  value,
  options,
  disabled = false,
  readOnly = false,
  noResultsMessage,
  onChange,
  buttonProps,
  dropdownTitle,
  errorMessage,
}) => {
  const { classes, cx } = useStyles();
  const { isTabletUp } = useScreenSize('Tablet');
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const countries = useMemo(() => options.slice(1), [options]);
  const [filteredOptions, setFilteredOptions] = useState(countries);
  const [selectedCountry, setSelectedCountry] = useState<Country | null>(
    () => countries.find((el) => el.value === value) || null
  );

  const onSearch = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value.toLowerCase();
    const sortedCountries = countries.sort((a, b) => {
      if (a.value < b.value) return -1;
      if (a.value > b.value) return 1;
      return 0;
    });
    const startsWith = sortedCountries.filter((country) => country.label.toLowerCase().startsWith(newValue));
    const contains = sortedCountries.filter((country) => country.label.toLowerCase().includes(newValue));
    setFilteredOptions([...new Set([...startsWith, ...contains])]);
  };

  const onClose = () => {
    setAnchorEl(null);
    setFilteredOptions(countries);
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    anchorEl ? onClose() : setAnchorEl(event.currentTarget);
  };

  const onSelectCountry = (country: Country) => {
    setSelectedCountry(country);
    onChange(country.value);
    onClose();
  };

  const renderFlag = (arrow: boolean) => {
    if (!selectedCountry?.value || !selectedCountry?.label) return <></>;
    return (
      <div className={classes.root}>
        <FlagIcon
          countryCode={selectedCountry.value}
          country={selectedCountry.label}
          arrow={arrow}
          disabled={disabled}
          className={classes.flagIcon}
        />
      </div>
    );
  };

  const renderSelect = () => {
    return (
      <>
        <div className={classes.root}>
          <ButtonBase
            {...buttonProps}
            color={errorMessage ? 'Error' : surface}
            disabled={disabled}
            className={cx(classes.select, { [classes.selectError]: !!errorMessage })}
            onClick={handleClick}
            focus="inset"
          >
            {renderFlag(true)}
          </ButtonBase>
        </div>
        {anchorEl && (
          <DropdownList
            anchor={anchorEl}
            onClose={onClose}
            onSearch={onSearch}
            searchAutoFocus
            inlineMessage={noResultsMessage}
            header={!isTabletUp && dropdownTitle}
            listItems={filteredOptions.map((language, index) => {
              return (
                <MenuItem
                  key={index}
                  label={language.label}
                  selected={language.value === selectedCountry?.value}
                  onClick={() => onSelectCountry(language)}
                  startIcon={<FlagIcon countryCode={language.value} country={language.label} />}
                />
              );
            })}
          />
        )}
      </>
    );
  };

  return <>{readOnly ? renderFlag(false) : renderSelect()}</>;
};

export default CountryDropdown;
