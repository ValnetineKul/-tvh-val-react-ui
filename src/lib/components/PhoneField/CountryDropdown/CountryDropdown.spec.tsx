import React from 'react';
import { render, screen, userEvent, initResponsiveTest } from '../../../test-utils';
import type { CountryCode } from '../../../types/common';
import CountryDropdown from './CountryDropdown';

const countries: { label: string; value: CountryCode }[] = [
  { label: 'Belgium', value: 'BE' },
  { label: 'Romania', value: 'RO' },
];

const props = {
  options: countries,
  value: 'BE' as CountryCode,
  dropdownTitle: 'Country',
};

describe('CountryDropdown', () => {
  it('Should render correctly', () => {
    const handleChange = jest.fn();
    const { container } = render(<CountryDropdown {...props} onChange={handleChange} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should call on change', () => {
    initResponsiveTest('Desktop');

    const handleChange = jest.fn();
    render(<CountryDropdown {...props} onChange={handleChange} />);

    const dropdown = screen.getByRole('button');
    userEvent.click(dropdown);

    expect(screen.getByRole('tooltip')).toBeInTheDocument();

    const albaniaButton = screen.getByRole('button', { name: 'Romania Romania' });
    userEvent.click(albaniaButton);

    expect(handleChange).toHaveBeenCalledWith('RO');
  });

  it('Should close when clicked again', () => {
    initResponsiveTest('Desktop');

    const handleChange = jest.fn();
    const { getByRole, queryByRole } = render(<CountryDropdown {...props} onChange={handleChange} />);

    const dropdown = getByRole('button');
    userEvent.click(dropdown);

    expect(getByRole('tooltip')).toBeInTheDocument();

    userEvent.click(dropdown);

    expect(queryByRole('tooltip')).not.toBeInTheDocument();
  });

  it('Should not render dropdown', () => {
    const handleChange = jest.fn();
    const { queryByRole } = render(<CountryDropdown {...props} onChange={handleChange} readOnly />);

    expect(queryByRole('button', { name: 'Belgium' })).not.toBeInTheDocument();
  });

  it('Should render the search results with starts with logic initially and then by contains logic and each part should be sorted alphabetically', () => {
    const options: { label: string; value: CountryCode }[] = [
      { label: 'Belgium', value: 'BE' },
      { label: 'Romania', value: 'RO' },
      { label: 'Nepal', value: 'NP' },
      { label: 'Japan', value: 'JP' },
      { label: 'Paraguay', value: 'PY' },
      { label: 'Panama', value: 'PA' },
    ];
    const handleChange = jest.fn();
    render(<CountryDropdown {...props} options={options} onChange={handleChange} />);

    const dropdown = screen.getByRole('button');
    userEvent.click(dropdown);

    const searchInput = screen.getByRole('searchbox');
    userEvent.type(searchInput, 'pa');

    const searchOptions = screen.getAllByRole('menuitem');
    expect(searchOptions).toHaveLength(4);
    expect(searchOptions[0]).toHaveTextContent('Panama');
    expect(searchOptions[1]).toHaveTextContent('Paraguay');
    expect(searchOptions[2]).toHaveTextContent('Japan');
    expect(searchOptions[3]).toHaveTextContent('Nepal');
  });
});
