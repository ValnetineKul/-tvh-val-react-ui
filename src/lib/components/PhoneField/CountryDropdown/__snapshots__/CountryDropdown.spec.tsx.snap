// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryDropdown Should render correctly 1`] = `
<div
  class="tss-1xhqyx3-CountryDropDown-root"
>
  <button
    class="MuiButtonBase-root tss-66scks-ButtonBase-root-ButtonBase-focusInset-CountryDropDown-select mui-10d1a0h-MuiButtonBase-root"
    tabindex="0"
    type="button"
  />
</div>
`;
